### PR TITLE
Add skiprows and nrows to parquet reader

### LINKED ISSUE #15144: [FEA] Add python bindings in the parquet reader for `num_rows`/`skiprows`
**Is your feature request related to a problem? Please describe.**
Unfortunately there has been churn in libcudf around support for `num_rows`/`skiprows` in the Parquet and ORC readers. In 22.08 we deprecated these parameters in the parquet reader (#11218) and then in 22.10 we removed them from C++ (#11503) and python (#11480). We also deprecated `num_rows`/`skiprows` in the ORC reader (#11522, see issue #11519).

At this point, we realized that chunked parquet reading (#11867) would require adding `num_rows`/`skiprows` back to the C++ implementation (#11657).

Let's stabilize row selection APIs in libcudf by completing these tasks:
- [ ] Add python bindings in the parquet reader for `num_rows`/`skiprows`
- [ ] Remove the deprecation notice in the ORC reader for `num_rows`/`skiprows` (#11522)

**Additional context**
We also dropped `num_rows`/`skiprows` support in the cuDF-python fuzz tests (#11505). My preference is to not include any python fuzz testing changes in the scope of this issue.

### PR DESCRIPTION
**Pull request by lithomas1**
## Description

closes #15144

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/io/parquet.py
Comment (wence-) [inline python/cudf/cudf/io/parquet.py:850]: No, I think this is wrong. These `dfs` are concatenated vertically, so after reading each df, one should do:
```
nrows = max(nrows - len(df), 0)
```

Updating `skip_rows` is more complicated because if you skipped all the rows in a file then you can't know if you should reduce `skip_rows` to zero or to `skip_rows - num_rows_in_file`.

Comment (wence-) [reply python/cudf/cudf/io/parquet.py:850]: Multiple files are, I believe, an implementation detail.

Comment (wence-) [reply python/cudf/cudf/io/parquet.py:850]: But let's cc @rjzamora to confirm

Comment (rjzamora) [reply python/cudf/cudf/io/parquet.py:850]: It seems a feels a bit ugly/inefficient to use `nrows` and/or `skiprows` when we are reading from a partitioned dataset, but you are right that we wouldn't want to pass these parameters "as provided" to `_read_parquet`.

The "efficient" solution would probably have us read the row-count metadata for each element of `key_paths` up front. This way we could avoid reading unnecessary files/rows altogether. Of course, if the user passes in a filter, we would need to read all data with the filter and perform the row-trimming after the fact.

>which we can't use anymore after pyarrow is gone

My understanding is that the pyarrow-removal effort only applies to the cython/c++ level. We are still allowed to depend on pyarrow at the python level (removing pyarrow would be a nightmare for dask-cudf at this point).

Comment (rjzamora) [reply python/cudf/cudf/io/parquet.py]: >At any rate, I think I might punt on this for now (since it doesn't look like making nrows/skiprows work for a partitioned dataset is high priority), and the rest of this PR is blocking nrows/skiprows support in the polars executor.

Yes - I don't see any reason to spend time optimizing nrows/skiprows for partitioned datasets. We can always revisit if necessary.

**Comment by mhaseeb123** [comment]
Hi @lithomas1, thank you for working on this PR. If not too much overhead, do you mind adding `num_rows` and `skip_rows` to the [ParquetReader](https://github.com/rapidsai/cudf/blob/3c83ce451446dfd556bd14ad8537b0189226a0e5/python/cudf/cudf/_lib/parquet.pyx#L807) as well and close #16249 as well

**Comment by lithomas1** [comment]
> Hi @lithomas1, thank you for working on this PR. If not too much overhead, do you mind adding `num_rows` and `skip_rows` to the [ParquetReader](https://github.com/rapidsai/cudf/blob/3c83ce451446dfd556bd14ad8537b0189226a0e5/python/cudf/cudf/_lib/parquet.pyx#L807) as well and close #16249

Sure, will give it a shot.

**Comment by lithomas1** [comment]
@mhaseeb123 

I'm hitting a bug in the chunked parquet reader for skip_rows > 0, so I don't think I can make further progress on this.
https://github.com/rapidsai/cudf/issues/16273

**Comment by mhaseeb123** [comment]
> @mhaseeb123
> 
> I'm hitting a bug in the chunked parquet reader for skip_rows > 0, so I don't think I can make further progress on this. #16273

If you are hitting this #16186, then please don't discard your changes. The bug should go away once #16195 merges. If you like you can pull in the changes from it and retest on your end  but we can wait until the merge as well to go ahead with this.

**Comment by lithomas1** [comment]
> Thanks for working on this. If you are hitting this #16186, then please don't discard your changes. The bug should go away once #16195 merges. If you like you can pull in the changes from it and retest on your end but we can wait until the merge as well to go ahead with this.

Thanks for the quick fix!

(I can't believe I missed that PR. The auto assigner assigned me to review that one too 😅 )

I currently have my changes stashed away in another branch, and I'll wait for your PR to land to merge that branch here.

File: python/cudf/cudf/tests/test_parquet.py
Comment (lithomas1) [inline python/cudf/cudf/tests/test_parquet.py:3799]: If you move the expected line before actual, these 3 tests should pass.

Not sure if this is a bug in libcudf or cudf-python, though.
(Looking at the buffers, they seem to be equal before and after the read_parquet_chunked call so maybe something strange is going on in cudf-python)

File: python/cudf/cudf/tests/test_parquet.py
Comment (lithomas1) [inline python/cudf/cudf/tests/test_parquet.py:3804]: I'm still getting an invalid device ordinal error here.

Comment (mhaseeb123) [reply python/cudf/cudf/tests/test_parquet.py:3804]: Working on this. Meanwhile, I noticed that if you read parquet with `cudf.read_parquet(numrows=.., skiprows=..)` the output columns look ok but the index column doesn't look right, hence you are getting a segfault when comparing the output with the original. Though, if you set `index=True` when `df.to_parquet()` then this goes away. It looks like the index column needs to be handled as well in Python/Cython layer when `index=None or False`.
```
# original table (nrows = 5, skip_rows=1)
    a    b
0   1   av
1   2   qw
2   3   hi
3   4  xyz
4   1   av
5   2   qw
6   3   hi
7   4  xyz
8   1   av
9   2   qw
10  3   hi
11  4  xyz
12  1   av
13  2   qw
14  3   hi
15  4  xyz
16  1   av
17  2   qw
18  3   hi
19  4  xyz

# got           
0         
1         
2         
3         
4         
5         
6         
7         
8         
9         
10        
11        
12        
13        
14  a    b
15  2   qw
16  3   hi
17  4  xyz
18  1   av
19  2   qw

#expected
   a    b
0  2   qw
1  3   hi
2  4  xyz
3  1   av
4  2   qw
```

Comment (mhaseeb123) [reply python/cudf/cudf/tests/test_parquet.py:3804]: Hi @galipremsagar @vyasr, if possible, could you please comment on if this is libcudf issue or if it should be handled in Cython/Python.

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_parquet.py:3804]: This needs to be handled in cython/python layer.

Comment (lithomas1) [reply python/cudf/cudf/tests/test_parquet.py:3804]: Thanks for finding this bug!

Looks like there is an issue processing of the range index metadata afterwards, taking a look!

File: python/cudf/cudf/tests/test_parquet.py
Comment (lithomas1) [reply python/cudf/cudf/tests/test_parquet.py:3799]: I can't reproduce this one anymore after fixing the bug in cudf.read_parquet with nrows/skip_rows.

I suppose the segfault corrupted something in the buffers?

File: python/cudf/cudf/tests/test_parquet.py
Comment (lithomas1) [reply python/cudf/cudf/tests/test_parquet.py:3801]: This should be fixed now!

Comment (mhaseeb123) [reply python/cudf/cudf/tests/test_parquet.py:3801]: @lithomas1  For the chunked parquet reader, unfortunately it looks like the segfault/invalid device ordinal problem isn't trivial so I would recommend not adding it to it until it's solved (in a follow up PR) and keep this PR moving.

Comment (mhaseeb123) [reply python/cudf/cudf/tests/test_parquet.py:3801]: Please open an issue to do that for now. Thank you!

File: python/cudf/cudf/tests/test_parquet.py
Comment (mhaseeb123) [inline python/cudf/cudf/tests/test_parquet.py:3821]:
```suggestion
def test_parquet_chunked_reader_nrows_skiprows(
    chunk_read_limit, pass_read_limit, nrows, skip_rows
):
    df = pd.DataFrame(
        {"a": [1, 2, 3, 4] * 100000, "b": ["av", "qw", "hi", "xyz"] * 100000}
    )
    expected = df[skip_rows:skip_rows+nrows]
    buffer = BytesIO()
    df.to_parquet(buffer)
    read = cudf.read_parquet(buffer, nrows=nrows, skip_rows=skip_rows)
    # Check results
    assert_eq(expected, read)
```

Comment (mhaseeb123) [inline python/cudf/cudf/tests/test_parquet.py:3821]:
```suggestion
def test_parquet_reader_nrows_skiprows(
    nrows, skip_rows
):
    df = pd.DataFrame(
        {"a": [1, 2, 3, 4] * 100000, "b": ["av", "qw", "hi", "xyz"] * 100000}
    )
    expected = df[skip_rows:skip_rows+nrows]
    buffer = BytesIO()
    df.to_parquet(buffer)
    got = cudf.read_parquet(buffer, nrows=nrows, skip_rows=skip_rows)
    # Check results
    assert_eq(expected, got)
```

File: python/cudf/cudf/tests/test_parquet.py
Comment (lithomas1) [reply python/cudf/cudf/tests/test_parquet.py:3801]: Let's keep track of this in #16195 - I added a comment there.

File: python/cudf/cudf/tests/test_parquet.py
Comment (lithomas1) [reply python/cudf/cudf/tests/test_parquet.py:3821]: If you want, I can keep this as a skipped test for you instead if you need the extra test parametrizations on the Python side.

Comment (mhaseeb123) [reply python/cudf/cudf/tests/test_parquet.py:3821]: No it's okay, we can remove it altogether for now or just test the normal reader. We can add some tests for the chunked reader when we have it working. Might be a few weeks.

**Comment by vyasr** [comment]
Is this PR blocked on resolving #16186 or is there a partial version that we want to land in the interim before that issue is fully resolved?

**Comment by mhaseeb123** [comment]
> Is this PR blocked on resolving #16186 or is there a partial version that we want to land in the interim before that issue is fully resolved?

Not blocked. We can merge this without adding bindings for `num_rows` and `skip_rows` to chunked PQ reader.

File: python/cudf/cudf/_lib/parquet.pyx
Comment (mhaseeb123) [inline python/cudf/cudf/_lib/parquet.pyx:380]: Let's be consistent with either `num_rows` or `nrows` across the files. @galipremsagar I can't find the same option in `pyarrow.read_table` or `pd.read_parquet` so am not sure which one to pick. If arbitrary, my vote would be `num_rows` to be consistent with C++ counterpart but not a blocker.

**Review by mhaseeb123 — CHANGES_REQUESTED** [review]
Just minor changes needed and should be good to go!

File: python/cudf/cudf/_lib/parquet.pyx
Comment (lithomas1) [reply python/cudf/cudf/_lib/parquet.pyx:380]: Yeah not sure which is better.

nrows would be consistent with read_csv, and num_rows would be consistent with libcudf.

Comment (mhaseeb123) [reply python/cudf/cudf/_lib/parquet.pyx:380]: Let's go with `nrows` then and further the PR to merge!

**Review by mhaseeb123 — APPROVED** [review]
As per conversation, please use `nrows` consistently on python side except when passing to libcudf. Looks good otherwise!

**Comment by mhaseeb123** [comment]
Some tests still failing with the following log. Looks like some more `nrows`/`num_rows` replacements needed.
```
FAILED io/test_parquet.py::test_read_parquet_basic[0-binary_source_or_sink1-nrows_skiprows0-columns1] - TypeError: read_parquet() got an unexpected keyword argument 'num_rows'
```

**Comment by mhaseeb123** [comment]
Note that the `nrows` only needs to be on Python side (read_parquet()) and the Cython layer may keep using `num_rows` if it's inline with other readers.

File: python/cudf/cudf/tests/test_parquet.py
Comment (wence-) [inline python/cudf/cudf/tests/test_parquet.py:1998]: *nit*: This comment doesn't align with the thing being tested.

File: python/cudf/cudf/tests/test_parquet.py
Comment (wence-) [inline python/cudf/cudf/tests/test_parquet.py:3809]:
```suggestion
    assert_eq(expected, got)
```

File: python/cudf/cudf/utils/ioutils.py
Comment (wence-) [inline python/cudf/cudf/utils/ioutils.py:206]: *nit (non-blocking)*: Not your fault, but there is no description anywhere (other than in the not very discoverable options documentation) of what low-memory mode means. So there's no way for someone to easily know what this means.

**Review by wence- — APPROVED** [review]
(Will need https://github.com/rapidsai/cudf/pull/16442 in to get past the cudf-polars test fails)

**Pull request merged**
